### PR TITLE
Add language to word timestamps for Whisper

### LINKED ISSUE #29520: Whisper language is not returned when using word timestamps
### System Info

- `transformers` version: 4.38.2
- Platform: macOS-13.6.1-x86_64-i386-64bit
- Python version: 3.10.11
- Huggingface_hub version: 0.21.3
- Safetensors version: 0.4.2
- Accelerate version: 0.27.2
- Accelerate config: 	not found
- PyTorch version (GPU?): 2.2.1 (False)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: no
- Using distributed or parallel set-up in script?: no

### Who can help?

@sanchit-gandhi @ArthurZucker

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

I am using the automatic speech recognition pipeline with whisper large v3. I want to get both the language as detected by whisper and the timestamps for the individual words. Both behaviors are supported individually( return_language=True, return_timestamps="word"), however when combined, the language is no longer returned.

This is the code i am using: 
```
import torch
from transformers import AutoModelForSpeechSeq2Seq, AutoProcessor, pipeline

model_id = "openai/whisper-large-v3"

model = AutoModelForSpeechSeq2Seq.from_pretrained(
    model_id, use_safetensors=True, torch_dtype=torch.float16, low_cpu_mem_usage=True,
)

processor = AutoProcessor.from_pretrained(model_id)

pipe = pipeline(
    "automatic-speech-recognition",
    model=model,
    tokenizer=processor.tokenizer,
    feature_extractor=processor.feature_extractor,
    max_new_tokens=128,
    chunk_length_s=30,
    batch_size=16,
    torch_dtype=torch.float16,
    return_language=True
)

result = pipe(audio_filepath, return_timestamps="word")
print(result)
```

I have done some digging in the source code and I believe I have found the problem in [tokenization_whisper.py](https://github.com/huggingface/transformers/blob/45c065109074d60c587d3e562f16531d02a422f6/src/transformers/models/whisper/tokenization_whisper.py#L1085)

Here, when using word timestamps, the existing chunks containing the language are ignored and only the words in the chunk are returned .
```
if return_timestamps or return_language:
        for chunk in chunks:
            if not return_timestamps:
                chunk.pop("timestamp")
            else:
                chunk["timestamp"] = tuple(chunk["timestamp"])
            if not return_language:
                chunk.pop("language")

        if return_timestamps == "word":
            new_chunks = []
            for chunk in chunks:
                new_chunks.extend(chunk["words"])
            optional = {"chunks": new_chunks}
        else:
            optional = {"chunks": chunks}
    else:
        optional = {}
    return full_text, optional
```

### Expected behavior

I expect to get both the language and the word timestamps when using `return_language=True` and `return_timestamps="word"`

Considering the source code mentioned above, I see 2 possible ways of achieving this.

1) Add an extra list to the chunks when using return_timestamps='word'

This would result in the format
```
{
  "text":"FULL TEXT", 
  "chunks": [
    "language": "LANGUAGE", 
    "timestamp": TIMESTAMP, 
    "text": "CHUNK_TEXT", 
    "words": [
      "text": "WORD", 
      "timestamp": WORD_TIMESTAMP
    ]
  ]
}
```

2) Add the language information to the words

This would maintain the current structure and simply add the language of the chunk to its words, giving the following format

```
{
  "text":"FULL TEXT", 
  "chunks": [
    "language": "LANGUAGE", 
    "timestamp": WORD_TIMESTAMP, 
    "text": "WORD"
  ]
}
```

I believe solution 1 could be achieved simply by removing the `if return_timestamps == "word":` statement completely.
I believe solution 2 could be achieved by replacing the `new_chunks.extend(chunk["words"])` with 
```
for word in chunk["words"]:
    word["language"] = chunk["language"]
    new_chunks.append(word)

### PR DESCRIPTION
**Pull request by algo-robin**
# Add language to word timestamps for Whisper

Fixes [29520](https://github.com/huggingface/transformers/issues/29520)

This fix enables whisper to return word-level timestamps and the predicted language at the same time. Before, enabling word-level timestamps would discard the language prediction.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?
@kamilakesbi 
@sanchit-gandhi

### CODE REVIEW

**Comment by kamilakesbi** [comment]
Hi @robinderat, thanks for working on this! 

Could you please add a test in `test_tokenization_whisper.py` to verify that we can indeed return both word-level timestamps and the predicted language at the same time?

**Comment by kamilakesbi** [comment]
LGTM! thanks for iterating on this :) 

cc @amyeroberts for final review!

**Review by sanchit-gandhi — APPROVED** [review]
Thanks for the contribution @robinderat! 🤗

**Comment by AvivSham** [comment]
@amyeroberts gentle reminder 😄

**Review by amyeroberts — COMMENTED** [review]
Thanks for the ping - just some questions to better understand the diff

File: tests/pipelines/test_pipelines_automatic_speech_recognition.py [hunk 1]
```diff
@@ -368,6 +367,63 @@ def test_return_timestamps_in_preprocess(self):
...
+        pipe = pipeline(
+            task="automatic-speech-recognition",
+            model="openai/whisper-tiny",
+            chunk_length_s=8,
+            stride_length_s=1,
+            return_language=True,
+        )
+        data = load_dataset("openslr/librispeech_asr", "clean", split="test", streaming=True, trust_remote_code=True)
+        sample = next(iter(data))
+        pipe.model.config.forced_decoder_ids = pipe.tokenizer.get_decoder_prompt_ids(language="en", task="transcribe")
```
[the remark this answers is not in the document]
Comment (robinderat) [reply tests/pipelines/test_pipelines_automatic_speech_recognition.py]: I based my test on an existing one that had this modification in it. However, it does not seem to affect the tests, so I have now removed them.

Specifying return_language=True in the pipeline is all that is required for this to work

**Review by amyeroberts — APPROVED** [review]
Thanks for adding and iterating on this!

**Pull request merged**